src/adaptors/adaptor_tls.h

```diff
@@ -0,0 +1,139 @@
+ * @param context - the context object
+ * @oaram conn_id - the connection id of the connection that creates the qd_tls_t
+ * @param log_source - the log source for TLS related logging.
+ */
+qd_tls_t *qd_tls(void *context, uint64_t conn_id, qd_log_source_t *log_source);
+
+/**
+ * Configure proton pn_tls objects using passed in config.
+ * and start the tls session by calling the pn_tls_start function.
+ * Searches for the sslProfile in the passed in config and loads the details from that sslProfile.
+ * Logs sslProfile configuration failures at INFO level.
+ *
+ * @param tls - Pointer to the qd_tls_t object which contains the tls session information.
+ * @param config   - Pointer to the qd_adaptor_config_t object which contains the sslProfile information.
+ * @param qd -     - Pointer to the qd_dispatch_t object.
+ * @param is_listener - set this flag to true if the tls session is initialized on an adaptor listener, false otherwise.
+ * @param alpn_protocols - An array of protocols supported by the application layer used when performing
+ * Application-Layer Protocol Negotiation (ALPN).
+ * @param alpn_protocol_count - The count of elements in the alpn_protocols array.
+ *
+ * @return
+ *   On success:
+ *     pn_tls_config and pn_tls_session objects in tls are set up.
+ *     Returns true
+ *   On failure:
+ *     Error log is written
+ *     All in-progress pn_tls objects are destroyed.
+ *     Returns false.
+ */
+bool qd_tls_start(qd_tls_t                  *tls,
+                  const qd_adaptor_config_t *config,
+                  const qd_dispatch_t       *qd,
+                  bool                       is_listener,
+                  const char                *alpn_protocols[],
+                  size_t                     alpn_protocol_count);
+
+/**
+ * Takes as many read buffers from the raw connection that is allowed by pn_tls_get_decrypt_input_buffer_capacity and
+ * decrypts those buffers and sticks the decrypted buffers in the passed in decrypted_buffs list.
+ *
+ * @param tls - The qd_tls_t object which contains the proton tls session information
+ * @param pn_raw_conn - pointer to the proton raw connection from which were going to take buffers from
+ * @param decrypted_buffs - A pointer to qd_adaptor_buffer_list which will contain the decrypted buffs. Make sure this
+ * list is initialized and empty.
+ *
+ * @return QD_TLS_ERROR if there is an error when trying to decrypt an encrypted buffer, otherwise return the number of
+ * undecrypted bytes processed.
+ *
+ */
+int qd_tls_decrypt(qd_tls_t *tls, pn_raw_connection_t *pn_raw_conn, qd_adaptor_buffer_list_t *decrypted_buffs);
+
+/**
+ * Encrypts a passed in single unencrypted qd_adaptor_buffer_t and populates the passed in qd_adaptor_buffer_list_t with
+ * the encrypted buffers. Can be called with a zero unencrypted_buff during the initiation of TLS handshake when the
+ * router is acting as the TLS client and needs to send out the initial hanshake frames.
+ *
```

Comment:
More correctly, may need to be called at any time with zero buffers if required by TLS protocol, not just during handshake.  I.e. any time you process input from peer, there is a chance that extra extra protocol bytes get generated regardless of the presence of application output.

Reply:
I have included this in the doc, thanks